### ISSUE TITLE
Leverage Service Worker and WebTransport on Public Gateways

### ISSUE THREAD

**Issue by lidel** [ipfs/in-web-browsers#207]
This is a stub/meta issue for tracking work to leverage Service Workers (SW) as a fallback on public gateways.

## Broad strokes idea

-  SW would intercept requests for specific content root (`/ipfs/cid/` or `/ipns/id`) and load data using
     -  Helia 
     - delegated routing over HTTPs (caskadht from cid.contact, `/routing/v1/providers/cid?cascade=ipfs-dht`)
     - bitswap over  WebTransport transport (allowing direct fetch from Kubo)
     
## When SW is registered / used

Below are high level ideas we may want to explore

### (A) Client-side recovery

When gateway returns error (list tbd, could be 429 and maybe also 502 503 and 504), and it knows request comes from web browser, return `text/html` response with option to recover via SW

This is a good candidate for dipping our toes, does not impact  "successful" flow, but allows for recovery from 429 Too Many Requests and Saturn CDN hiccups we will see in the near future.
     
### (B) Ability to register global worker 

Giving user ability to prefer local JS for future requests, not only for errors, but all responses.

### (C) Ability to isolate path gateway 

We could do something related to https://github.com/ipfs/in-web-browsers/issues/157 where actual payload is loaded in [sandboxed iframe](https://web.dev/sandboxed-iframes/)

**Comment by lidel** [comment]
> [..] The gateway needs to be loaded explicitly which can call navigator.serviceWorker.register().

Yes, just like you noted, the idea is the gateway will return HTML that loads JS responsible for registering SW.

Initially, we would be doing that on HTTP 502 and 504 error pages (show a button, which when clicked, registers worker, and  reloads page). This allows us to figure out SW parts without having to implement full gateway (we only implement flat files).

> I feel the users should be oblivious to the fact where the content is being loaded from. This should be local first by-default and recover from a public gateway if needed. [..] loads the gateway in an iframe [..]

Yes, that would be the endgame for same origin requests send by web browser for the root document.
But I urge everyone to NOT focus on that, and instead make HTTP error recovery UX pleasant and useful first.

Rationale: start with smaller scope, make things robust first,  Reimplementing the [full gateway spec](https://github.com/ipfs/specs/tree/main/http-gateways#readme) in JS to support for anything other than flat deserialized files will be a lot of work, and we need to have a reliable way of testing ([gateway tests are wip](https://github.com/ipfs/gateway-conformance)).

**Comment by SgtPooki** [comment]
## 2023-03-17 Discussion in weekly standup

* serviceworker should be able to take request that was going to service worker and send somewhere else.. 
* not hosting on gateway to start, simple html page that can be rendered locally in browser.

### End Goal

* 


### basic plan
* simple html page that can do content routing: accept URL/CID in input element, return cat jpg beneath it (result of `IPFS.get(URL/CID)`)
* accept input element info via URL to the html page, should auto-fill input element and submit; render cat jpg beneath it
   * Then support video rendering?
* migrate code to library and demo to using service-worker
* service-worker is pluggable for different retrieval methods (bitswap, libp2p stream, libp2p graph...?)

### potential showoff items
* Implement a brand new (madeup) transport and implement in demo - show users they can do this without going through bureaucracy. 
* Allow graph based requests over http/whatever

### workstreams:

* Demo of app that fetch blocks over http/bitswap from providers (direct).. verify blocks, render content
* Use same logic, but over libp2p stream.

### initial action items

- [ ] @aschmahmann to provide basic diagram of "IPFS components" necessary to make this work
- [ ] @SgtPooki to build demo repo, html page and basic layout that we can start working in.
- [ ] @MarcoPolo to build "fetch over libp2p"

### open questions

- [ ] can service worker intercept Libp2p requests? (foreign-fetch).
   - Can we implement browser hash URL standard for translating a local request to libp2p so service worker intercepts it (http://mysite.com/ipfsBrowser#libp2p:stuff/here -> intercepted by service worker, does the things)


### callouts
* will bundle with webpack and build locally runnable site (npm start) at first. webpack should help building a single "deployable" html page that can simply be ran locally.
